### PR TITLE
Add trapFocus function

### PR DESCRIPTION
**Pull request by schoenkaft**
Use these functions a lot in modals/dialogs, so I guess it's not too random in a library like this, right? 

- You can quickly disable all elements expect the one your trying to 'single-focus'. 
- It will not affect any existing `tabindex="-1"` elements and it will clean up afterwards.
- It adds a property to the node object, which can feel a bit arbitrary. It felt a bit safer than adding a data-attribute (e.g. something like `data-focus-disabled="true"`, which feels more prone to errors). Better ideas?

Feedback is welcome!

PS: tried to simulate keyboard focus events, but didn't work out 😞

### CODE REVIEW

File: functions/disableDocumentFocus.mjs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+/* eslint-disable no-param-reassign */
+
+/**
+ * Disable focus for nodes other than specified.
+ */
+const disableDocumentFocus = focusNode => {
+  const nodes = [...document.querySelectorAll(`:not([tabindex="-1"])`)]
```
[the remark this answers is not in the document]
Comment (harmenjanssen) [reply functions/disableDocumentFocus.mjs]: Also, I would `document` to be a parameter (`rootNode = document` with a default would be great).

[the remark this answers is not in the document]
Comment (schoenkaft) [reply functions/disableDocumentFocus.mjs]: > Might be a good idea for performance to restrict this query to the limited set of focusable elements? (`a`, `button`, `input`, etc)

Kind of makes sense, although the DOM is super fast and this shouldn't trigger any reflow. You'll have to be quite thorough and specific in what to target... this im my list for now:

```js
const CANDIDATES = `
  a, button, input, select, textarea, area, iframe, object, embed,
  [tabindex], [contenteditable]
`;
```

**Comment by schoenkaft** [comment]
> [...]
> This has the added benefit that your `enable` function has a reference to that original scope, so you don't have to query the nodes again, and you can even skip the custom property.
> 
> What do you think?

Yeah, makes a lot of sense! Some initial thoughts / changes:

- Let's ditch `enableDocumentFocus` and  🤘
- Calling it `disableDocumentFocus` doesn't feel right anymore, since you can also specify the `rootNode` now. I came up with `trapFocus`... thoughts?
- Kind of makes sense to 'disable' when calling `trapFocus`, but also kind of makes sense to do it via it's API since you'll also 'enable' that way? So `trapFocus.enable()` and `trapFocus.disable()`
- When calling it `trapFocus`, `enable`/`disable` might mean different things based or if you would call the constant something like: `const disabledDocument = disableDocumentFocus(modalDialog);`. So thoughts on `enabled/disable`, `trap/untrap`, `activate/deactivate`, ...?

Thoughts on the above are welcome, and also if we should initially trap or should 'enable/disable' both via an exposed method?

```js
const focusTrap = trapFocus(modal);
focusTrap.untrap();

// vs...

const focusTrap = trapFocus(modal);
focusTrap.activate();
focusTrap.deactivate();
```

**Comment by schoenkaft** [comment]
I've updated the test and came up with this for now:

```js
const focusTrap = trapFocus(modal);
focusTrap.activate(); // activates the trap
focusTrap.deactivate(); // deactivates the trap
focusTrap.nodes; //=> array of nodes
```

File: functions/trapFocus.mjs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  const activate = () => nodes.forEach(node => node.setAttribute('tabindex', '-1'));
+  const deactivate = () => nodes.forEach(node => node.removeAttribute('tabindex'));
+
+  return {
+    nodes,
+    activate,
+    deactivate,
+  };
+};
```
Comment (harmenjanssen) [inline functions/trapFocus.mjs]: I love the simplicity of this. 😍

**Comment by harmenjanssen** [comment]
As discussed: I'm really fond of `release` as opposite of `trap`, mostly because I think it's an elegant wording and a less generic verb than `(de)activate`. 🎓

I also think there's one more way we could reason about this:

```js
const releaseFocus = trapFocus(modal);
releaseFocus();
```

In which `trapFocus` does not return an object, but the reversing function directly. (you could still support a `.nodes` property on the function, which I like)

I'm not sure which is better, in summary:

```js
// 1
const modalFocusTrap = trapFocus(modal);
modalFocusTrap.active();
modalFocusTrap.deactive();

// 2
const modalFocusTrap = trapFocus(modal);
modalFocusTrap.release();

// 3
const releaseModalFocusTrap = trapFocus(modal);
releaseModalFocusTrap();
```

@martijngastkemper I'm really curious what you think. I personally dislike number **1**, but don't have a strong preference for **2** or **3**.
@schoenkaft Do you have feelings for **3**?

**Comment by schoenkaft** [comment]
> @schoenkaft Do you have feelings for 3?

Yep! Although I like the idea, it feels really unintuitive to me. I'd say let's go for 2 then. And example 1 would be `focusTrap` (with `focusTrap.activate()`) instead of `trapFocus` (and `trapFocus.active()`) I'd say, but I think 2 is more in line with functions already in the library.

**Comment by harmenjanssen** [comment]
Sounds good!

**Comment by schoenkaft** [comment]
I did expose the `trap` function as well, but it actually doesn't make a lot sense to call that later on if the initial function already traps it. Right?

So:

```js
const focusTrap = trapFocus(modal);
focusTrap.release(); 

// but no focusTrap.trap() to trap it again... might make it to complicated
```

PS: but still curious what @martijngastkemper and @MicheleNL think though 😄

**Comment by martijngastkemper** [comment]
I like option 2. Option 1 does the job, but in the context of element focus 2 is more intuitive. When quickly scanning the code I can imagine dev's think that activate applies to the modal.

Option 3: I don't expect `trapFocus` to return a function. I expect an object. And it can't handle `retrap`, if that feature is useful.

**Comment by schoenkaft** [comment]
So it's going to be:

```js
const focusTrap = trapFocus(modal); // traps
focusTrap.release(); // releases
focusTrap.retrap(); // retraps
```

**Pull request merged**